### PR TITLE
`SubspaceDiscrete.from_simplex` convenience constructor

### PR DESCRIPTION
**Pull request by AdrianSosic**
This PR adds a new `SubspaceDiscrete.from_simplex` convenience constructor that is useful for mixture use cases. In particular, it enables such cases when an API is involved, as the two existing alternatives will not work here:
* `from_product` in combination with sum constraints suffers from the exponential blowup of parameter configurations
* `from_dataframe` requires sending excessively large requests to the API (since the requests need to contain the full enumerated search space)

### CODE REVIEW

File: baybe/searchspace/discrete.py [hunk 1]
```diff
@@ -230,6 +230,85 @@ def discrete_parameter_factory(
...
+                    (row_sums < total - tolerance) | (row_sums > total + tolerance)
+                ].index
+            else:
+                rows_to_drop = row_sums[row_sums > total + tolerance].index
+            df.drop(rows_to_drop, inplace=True)
+
+        # Incrementally build up the space, dropping invalid configuration along the way
+        exp_rep = pd.DataFrame({parameters[0].name: parameters[0].values})
+        drop_invalid(exp_rep, equality=False)
+        for param in parameters[1:]:
```
Comment (AVHopp) [inline baybe/searchspace/discrete.py:297]: So what I get from here is that you iteratively add "the next" parameter and then filter out those tows whose sum now becomes too large, right?
This in particular means that (i) the performance of this method depends on the order of the parameters and (ii) that this might be inefficient, depending on the order of the parameters (imagine only one parameter being respondible for dropping rows, and that this happens to be the last).

The least I'd do is to put some information about this issue in the docstring of the function. Ideally, one could adjust the code such that these issues cannot occur at all. This should be in theory possible (or at least improvable) by e.g. re-ordering parameters or looking at specific algorithms for the SubsetSum problem.

HOWEVER, I know that this might not even be possible or a huge amount of additional work. I will thus also accept the current version but wanted to spark the discussion on that and hear your opinion :)

File: tests/hypothesis_strategies/alternative_creation/test_searchspace.py [hunk 2]
```diff
@@ -97,3 +101,34 @@ def test_searchspace_creation_from_dataframe(df, parameters, expected):
     else:
         with pytest.raises(expected):
             SearchSpace.from_dataframe(df, parameters)
+
+
+@given(
+    parameters=st.lists(
+        numerical_discrete_parameter(min_value=0.1, max_value=1.0),
```
Comment (AVHopp) [inline tests/hypothesis_strategies/alternative_creation/test_searchspace.py:108]: Why not allowing min_value=0.0?

File: tests/hypothesis_strategies/alternative_creation/test_searchspace.py [hunk 3]
```diff
@@ -97,3 +101,34 @@ def test_searchspace_creation_from_dataframe(df, parameters, expected):
...
+        parameters, total=total, boundary_only=False, tolerance=tolerance
+    )
+    assert (subspace.exp_rep.sum(axis=1) <= total + tolerance).all()
+
+
+def tests_discrete_space_creation_from_simplex_boundary():
+    """Candidates from a simplex boundary space satisfy the boundary constraint."""
+    total = 1.0
+    tolerance = 1e-6
+    parameters = [
```
Comment (AVHopp) [inline tests/hypothesis_strategies/alternative_creation/test_searchspace.py:127]: Why are these not provided via @given?

**Review by AVHopp — COMMENTED** [review]
Some minor stuff and one bigger that we might want to discuss. But I already like it very much and there are no real blocking points from my end

File: CHANGELOG.md [hunk 4]
```diff
@@ -10,6 +10,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - `mypy` for campaign, constraints and telemetry
 - Top-level example summaries
 - `RecommenderProtocol` as common interface for `Strategy` and `Recommender`
+- `SubspaceDiscrete.from_simplex` convenience constructor
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply CHANGELOG.md:13]: Done, thx (there was only the "discrete" missing, in fact)

File: baybe/searchspace/discrete.py [hunk 5]
```diff
@@ -230,6 +230,85 @@ def discrete_parameter_factory(
...
 
+    @classmethod
+    def from_simplex(
+        cls,
+        parameters: List[NumericalDiscreteParameter],
+        total: float,
+        boundary_only: bool = False,
+        tolerance: float = 1e-6,
+    ) -> SubspaceDiscrete:
+        """Efficiently create discrete simplex subspaces.
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/searchspace/discrete.py:241]: Yup, no problems here

File: baybe/searchspace/discrete.py [hunk 6]
```diff
@@ -230,6 +230,85 @@ def discrete_parameter_factory(
...
+        # Validate parameter types
+        if not (all(isinstance(p, NumericalDiscreteParameter) for p in parameters)):
+            raise ValueError(
+                f"All parameters passed to '{cls.from_simplex.__name__}' "
+                f"must be of type '{NumericalDiscreteParameter.__name__}'."
+            )
+
+        # Validate non-negativity
+        min_values = [min(p.values) for p in parameters]
+        if not (min(min_values) >= 0.0):
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/searchspace/discrete.py:277]: I guess theoretically one could also extend this to arbitrary values, but I restricted it for now for the following reasons:
* The standard definition of a "simplex" is indeed for non-negative values only!
* The motivation for the method were the mixture use cases we already have, and there you'll always have non-negativity
* Most critically, the shortcut I took for the creation would be much more complicated if you also allowed negative values, because then you can still add parameters and reduce the total sum. 

So mostly for the last reason, this is omitted. I'm open to extend the method in the future, but this would definitely be a second PR and I'd only do it if we have a use case that needs it (otherwise, we have more important issues to solve 😄 )

If you are OK with that, please resolve

File: baybe/searchspace/discrete.py (hunk 1 above)
Comment (AdrianSosic) [reply baybe/searchspace/discrete.py:297]: True, the efficiency gains depend on the specific order of your parameter sequence. I am happy to add a note to the docstring mentioning this fact 👍🏼 But I also think (as you already said) that optimizing this part is a lot of work though there is (at least currently) no real use case that would benefit from it, so let's delay that until really needed. Nonetheless, appreciate your comment!

And BTW: I actually forgot to implement a slight improvement, and will do it now: instead of throwing out only configurations that exceed the sum at the current moment, we can additionally keep track of the minimum contribution that will come from the yet-to-be-added parameters and apply a stricter threshold that will be adjusted after each added parameter. Will ping you once it's there

File: baybe/searchspace/discrete.py [hunk 7]
```diff
@@ -230,6 +230,85 @@ def discrete_parameter_factory(
...
+        tolerance: float = 1e-6,
+    ) -> SubspaceDiscrete:
+        """Efficiently create discrete simplex subspaces.
+
+        The same result can be achieved using
+        :meth:`baybe.searchspace.discrete.from_product` in combination with appropriate
+        sum constraints. However, such an approach is inefficient because the Cartesian
+        product involved creates an exponentially large set of candidates, most of
+        which do not satisfy the simplex constraints and must be subsequently be
+        filtered out by the method.
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/searchspace/discrete.py]: you mean "satisfy the simplex constraint" instead of "satisfy the simplex constraints"?

[the remark this answers is not in the document]
Comment (AVHopp) [reply baybe/searchspace/discrete.py]: Ahhhh I once again misread. Everything gucci here :)

File: baybe/searchspace/discrete.py [hunk 8]
```diff
@@ -230,6 +230,85 @@ def discrete_parameter_factory(
...
+
+        By contrast, this method uses a shortcut that removes invalid candidates
+        already during the creation of parameter combinations, resulting in a
+        significantly faster construction.
+
+        Args:
+            parameters: The parameter to be used for the simplex construction.
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/searchspace/discrete.py]: fixed, thanks

File: baybe/searchspace/discrete.py [hunk 9]
```diff
@@ -230,6 +231,99 @@ def discrete_parameter_factory(
 
         return cls(parameters=parameters, exp_rep=df, empty_encoding=empty_encoding)
 
+    @classmethod
+    def from_simplex(
+        cls,
+        parameters: List[NumericalDiscreteParameter],
+        total: float,
```
Comment (AdrianSosic) [inline baybe/searchspace/discrete.py:238]: @AVHopp @Scienfitz: what do you think about the names of the kw-args `total` and `boundary_only`. Could not come up with better ideas, perhaps you have some.... Especially `total` could be a bit confusing, because if `boundary_only=False`, the parameter sum can also be smaller that `total`.

File: tests/hypothesis_strategies/alternative_creation/test_searchspace.py (hunk 2 above)
Comment (AdrianSosic) [reply tests/hypothesis_strategies/alternative_creation/test_searchspace.py:108]: Mostly just to limit the test case a bit. You see, if you put 0.0 instead, then the simplex can contain an arbitrary number of parameters, because each parameter can just contribute 0.0 to the sum (provided is has 0.0 as a value). But putting 0.1, we know that the simplex can span at most 10 parameters, due to the total being 1.0. The alternative – which would work just as well – is to limit the number of parameter in the list created by hypothesis. I'd be fine with both solutions. Which one do you prefer? Or do you have an even better idea?

File: tests/hypothesis_strategies/alternative_creation/test_searchspace.py (hunk 3 above)
Comment (AdrianSosic) [reply tests/hypothesis_strategies/alternative_creation/test_searchspace.py:127]: Just so that we are on the same page: you suggest that these should be randomly generated by hypothesis as well, correct? In principle I have nothing against that, but I think it is really hard to find a setting that does produce **useful** tests if we did that – which was the reason why I used manual parameterization here. The problem is that you can very easily end up with cases where not a single parameter configuration falls into the requested simplex, in which case the test would trivially succeed (since it checks if the returned configurations, which would be empty then, satisfy the constraint). That is, you need to ensure that at least some of the randomly generated parameters can actually combined such taht you exactly hit the simplex boundary.

Example that can easily happen when you randomly generate without placing very precise restrictions:
* Param1: [0.0, 0.2, 0.5]
* Param2: [0.0, 0.1]
* requested total: 1.0

--> nothing lands on the simplex boundary, test succeeds trivially

You see what I mean? Any ideas?

File: baybe/searchspace/discrete.py (hunk 1 above)
Comment (AdrianSosic) [reply baybe/searchspace/discrete.py]: And the note is now there as well. Let me know if OK

File: baybe/searchspace/discrete.py [hunk 10]
```diff
@@ -230,6 +230,85 @@ def discrete_parameter_factory(
...
+
+        # Validate non-negativity
+        min_values = [min(p.values) for p in parameters]
+        if not (min(min_values) >= 0.0):
+            raise ValueError(
+                f"All parameters passed to '{cls.from_simplex.__name__}' "
+                f"must have non-negative values only."
+            )
+
+        def drop_invalid(df: pd.DataFrame, equality: bool) -> None:
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/searchspace/discrete.py]: Have expanded the docstring (and renamed the method parameters). Please resolve if happy

File: baybe/searchspace/discrete.py (hunk 1 above)
Comment (AdrianSosic) [reply baybe/searchspace/discrete.py]: .... aaaaand the extension for arbitrary discrete parameter is not there as well. Done. Happy reviewing 🙃

File: tests/hypothesis_strategies/alternative_creation/test_searchspace.py (hunk 2 above)
Comment (AVHopp) [reply tests/hypothesis_strategies/alternative_creation/test_searchspace.py:109]: Ok, I get the point. I'd prefer the variant where we explicitly limit the number of parameters, as the case of a parameter having a value of 0 seems like something we should test.
Did you actually test of what you are worried about actually happens/makes any difference?

File: tests/hypothesis_strategies/alternative_creation/test_searchspace.py (hunk 3 above)
Comment (AVHopp) [reply tests/hypothesis_strategies/alternative_creation/test_searchspace.py:128]: To be honest, I would have expected that hypothesis takes care of that. Isn't the whole point of us using it that we do not need to have manual lists at all? I would at least then include an additional test with hypothesis generating the parameters on its own, then we are certain that we test both important edge cases and the cases hypothesis might find.

File: baybe/searchspace/discrete.py [hunk 11]
```diff
@@ -230,6 +231,122 @@ def discrete_parameter_factory(
...
+        for i, (param, min_to_go) in enumerate(
+            zip_longest(numerical_parameters, min_upcoming, fillvalue=0)
+        ):
+            if i == 0:
+                exp_rep = pd.DataFrame({param.name: param.values})
+            else:
+                exp_rep = pd.merge(
+                    exp_rep, pd.DataFrame({param.name: param.values}), how="cross"
+                )
+            drop_invalid(exp_rep, total - min_to_go, boundary_only=False)
```
Comment (AVHopp) [inline baybe/searchspace/discrete.py:334]: The way that you use `drop_invalid` here does not fit its docstring, since the second item is not `total` anymore.

File: baybe/searchspace/discrete.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (AVHopp) [reply baybe/searchspace/discrete.py]: In general I am happy :) Just double-check that the description of `total` is accurate w.r.t. your usage of this function. If this is the case, please resolve.

File: baybe/searchspace/discrete.py (hunk 1 above)
Comment (AVHopp) [reply baybe/searchspace/discrete.py]: Looks good in general :) The parts that I have questions about are marked via comments, hence resolving this one.

File: baybe/searchspace/discrete.py (hunk 9 above)
Comment (Scienfitz) [reply baybe/searchspace/discrete.py:238]: fine with both but indeed the `max` version might be better due to the mentioned case

Comment (AdrianSosic) [reply baybe/searchspace/discrete.py:238]: Problem with `max` is that it's a python keyword. So `max_sum` then?

File: tests/hypothesis_strategies/alternative_creation/test_searchspace.py (hunk 2 above)
Comment (Scienfitz) [reply tests/hypothesis_strategies/alternative_creation/test_searchspace.py:109]: agree with alex, limits via min value seems like a detour to achieve whats possible also by setting the corresponding entry in the draw AND having 0.0 in there seems important to me as this will happen for lots of use cases

File: tests/hypothesis_strategies/parameters.py [hunk 12]
```diff
@@ -2,6 +2,7 @@
 
 import hypothesis.strategies as st
 import numpy as np
+from altair import Optional
```
Comment (Scienfitz) [inline tests/hypothesis_strategies/parameters.py:5]: dont know this can you explain? any chance the autoimport went wrong and should use typing?

File: tests/hypothesis_strategies/alternative_creation/test_searchspace.py (hunk 3 above)
Comment (Scienfitz) [reply tests/hypothesis_strategies/alternative_creation/test_searchspace.py:128]: why not let it draw random parameters and extract from them the `total` instead of fixing it here, eg as sum of max of all values

File: tests/hypothesis_strategies/parameters.py [hunk 13]
```diff
@@ -75,14 +76,20 @@ def custom_descriptors(draw: st.DrawFn):
...
 def numerical_discrete_parameter(
     draw: st.DrawFn,
+    min_value: Optional[float] = None,
+    max_value: Optional[float] = None,
 ):
     """Generate :class:`baybe.parameters.numerical.NumericalDiscreteParameter`."""
     name = draw(parameter_name)
     values = draw(
         st.lists(
             st.one_of(
```
Comment (Scienfitz) [inline tests/hypothesis_strategies/parameters.py:86]: is the `one_of` still required since you kicked ints?

**Review by Scienfitz — APPROVED** [review]
very nice, could set us up for an upcoming ecosystem of searchspace combiner operations and more convenience contructors

File: baybe/searchspace/discrete.py (hunk 9 above)
Comment (Scienfitz) [reply baybe/searchspace/discrete.py:238]: (yes that is what I meant just like alex)

Comment (AdrianSosic) [reply baybe/searchspace/discrete.py]: Alright, fixed

File: baybe/searchspace/discrete.py [hunk 14]
```diff
@@ -230,6 +231,122 @@ def discrete_parameter_factory(
...
+        subsequently be filtered out by the method.
+
+        By contrast, this method uses a shortcut that removes invalid candidates
+        already during the creation of parameter combinations, resulting in a
+        significantly faster construction.
+
+        Args:
+            parameters: The parameters to be used for building the subspace.
+                Parameters of type
+                :class:`baybe.parameters.numerical.NumericalDiscreteParameter`
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/searchspace/discrete.py]: Good idea 💡 Changed it accordingly

File: baybe/searchspace/discrete.py [hunk 15]
```diff
@@ -230,6 +231,122 @@ def discrete_parameter_factory(
...
+        Returns:
+            The created simplex subspace.
+
+        Note:
+            The achieved efficiency gains can vary depending on the particular order in
+            which the parameters are passed to this method, as the configuration space
+            is built up incrementally from the parameter sequence.
+        """
+        # Separate parameter into numerical and other
+        numerical_parameters = [
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/searchspace/discrete.py]: jup, done

File: tests/hypothesis_strategies/parameters.py (hunk 12 above)
Comment (AdrianSosic) [reply tests/hypothesis_strategies/parameters.py:5]: ahem .... just me getting accustomed to using VScode 🤣  thanks for the catch, fixed it

File: baybe/searchspace/discrete.py [hunk 16]
```diff
@@ -230,6 +231,122 @@ def discrete_parameter_factory(
...
+
+            Args:
+                df: The dataframe whose rows should satisfy the simplex constraint.
+                total: The maximum row sum defining the simplex size.
+                boundary_only: Flag to control if the points represented by the rows
+                    may lie inside the simplex or on its boundary only.
+            """
+            row_sums = df.sum(axis=1)
+            if boundary_only:
+                rows_to_drop = row_sums[
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/searchspace/discrete.py:318]: Good point. Made it even more explicit by calling it `locs_to_drop` because all "idx"-based naming is also always somewhat ambiguous in dataframes (dammit pandas!!!)

File: tests/hypothesis_strategies/parameters.py (hunk 13 above)
Comment (AdrianSosic) [reply tests/hypothesis_strategies/parameters.py:87]: Nope, excellent catch, thanks! Fixed

File: tests/hypothesis_strategies/alternative_creation/test_searchspace.py (hunk 3 above)
Comment (AdrianSosic) [reply tests/hypothesis_strategies/alternative_creation/test_searchspace.py]: Yeah, refactored the stuff a bit and went back a bit closer to what I had at the very beginning (one single parametrized test for both `boundary_only` options). I think the current version should do it's job and satisfies your request

File: tests/hypothesis_strategies/alternative_creation/test_searchspace.py (hunk 2 above)
Comment (AdrianSosic) [reply tests/hypothesis_strategies/alternative_creation/test_searchspace.py]: Ok, valid point. Changed it accordingly 👍🏼 Now restricting the number of parameters instead

File: baybe/searchspace/discrete.py (hunk 11 above)
Comment (AdrianSosic) [reply baybe/searchspace/discrete.py]: Can you elaborate what exactly you mean? Note btw that the "total" defined in this function locally does NOT need to coincide with the "total" defined in the outer scope. Therefore, it still DOES implement a simplex constraint, though a different simplex than the outside one. Does this resolve the issue? Anyway, could you just have a look at the latest version and let me know if you still think it's an issue?

Comment (AVHopp) [reply baybe/searchspace/discrete.py]: Ok, with the interpretation of this being technically a different simplex this makes sense.

**Review by AVHopp — APPROVED** [review]
